File: Criacionais/Prototype e Singleton/Exercício/exercicio.py
# Classe Aluno (Prototype)
class Aluno:
    def __init__(self, nome, idade, curso, unidade, periodo):
        self.nome = nome
        self.idade = idade
        self.curso = curso
        self.unidade = unidade
        self.periodo = periodo

    # Clone (cópia superficial)
    def clone(self):
        return Aluno(
            self.nome,
            self.idade,
            self.curso,
            self.unidade,
            self.periodo
        )


# Singleton - GerenciadorAlunos
class GerenciadorAlunos:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(GerenciadorAlunos, cls).__new__(cls)
            cls._instance.lista_matriculados = []
        return cls._instance

    def adicionar_aluno(self, aluno):
        self.lista_matriculados.append(aluno)
        print(f"> Sucesso: {aluno.nome} matriculado em {aluno.curso}.")

    def listar_matriculados(self):
        print("\nLISTA DE MATRÍCULAS")
        if not self.lista_matriculados:
            print("Nenhum aluno cadastrado.")
        else:
            for aluno in self.lista_matriculados:
                print(vars(aluno))


# Client
sigaA = GerenciadorAlunos()
sigaB = GerenciadorAlunos()

print(f"Os gerenciadores são a mesma instância? {'SIM' if sigaA is sigaB else 'NÃO'}")

protoADS = Aluno("Padrão", 0, "Análise de Sistemas", "Unidade Centro", "Noite")

joao = protoADS.clone()
joao.nome = "João da Silva"
joao.idade = 20

maria = protoADS.clone()
maria.nome = "Maria Souza"
maria.idade = 22

sigaA.adicionar_aluno(joao)
sigaA.adicionar_aluno(maria)

sigaB.listar_matriculados()

print("\nVerificação de Integridade do Protótipo:")
print(f"Nome no Protótipo: {protoADS.nome}")
print(f"Idade no Protótipo: {protoADS.idade}")

File: Criacionais/Prototype e Singleton/Exercício/exercicio.js
// Classe Aluno que será clonada
class Aluno {
    constructor(nome, idade, curso, unidade, periodo) {
        this.nome = nome;
        this.idade = idade;
        this.curso = curso;
        this.unidade = unidade;
        this.periodo = periodo;
    }

    // Método para clonar o objeto atual (Cópia Superficial)
    clone() {
        return new Aluno(
            this.nome, 
            this.idade, 
            this.curso, 
            this.unidade, 
            this.periodo
        );
    }
}

// SIGA
class GerenciadorAlunos {
    constructor() {
        if (GerenciadorAlunos.instance) {
            return GerenciadorAlunos.instance;
        }

        this.listaMatriculados = [];
        GerenciadorAlunos.instance = this;
    }

    adicionarAluno(aluno) {
        this.listaMatriculados.push(aluno);
        console.log(`> Sucesso: ${aluno.nome} matriculado em ${aluno.curso}.`);
    }

    listarMatriculados() {
        console.log("\nLISTA DE MATRÍCULAS");
        if (this.listaMatriculados.length === 0) {
            console.log("Nenhum aluno cadastrado.");
        } else {
            console.table(this.listaMatriculados);
        }
    }
}

// Client

const sigaA = new GerenciadorAlunos();
const sigaB = new GerenciadorAlunos();

console.log(`Os gerenciadores são a mesma instância? ${sigaA === sigaB ? "SIM" : "NÃO"}`);

const protoADS = new Aluno("Padrão", 0, "Análise de Sistemas", "Unidade Centro", "Noite");

const joao = protoADS.clone();
joao.nome = "João da Silva";
joao.idade = 20;

const maria = protoADS.clone();
maria.nome = "Maria Souza";
maria.idade = 22;

sigaA.adicionarAluno(joao);
sigaA.adicionarAluno(maria);

sigaB.listarMatriculados();

console.log("\nVerificação de Integridade do Protótipo:");
console.log(`Nome no Protótipo: ${protoADS.nome}`);
console.log(`Idade no Protótipo: ${protoADS.idade}`);
